File: Media_tracker_web/src/php/models/Last.FM/LastFmAlbum.php
<?php

class LastFmAlbum {
    public string $name;
    public string $artist;
    public ?string $mbid;
    public string $url;
    public ?string $releaseDate;
    public int $playCount;
    public int $listeners;
    public ?string $imageUrl;
    public array $tags = [];
    public array $tracks = [];

    public function __construct(array $json) {
        $this->name = $json['name'] ?? '';
        $this->artist = $json['artist']['name'] ?? '';
        $this->mbid = $json['mbid'] ?? null;
        $this->url = $json['url'] ?? '';
        $this->releaseDate = trim($json['releasedate'] ?? '');
        $this->playCount = (int) ($json['playcount'] ?? 0);
        $this->listeners = (int) ($json['listeners'] ?? 0);

        // Grab largest image available
        $imageList = array_filter($json['image'] ?? [], fn($img) => !empty($img['#text']));
        $image = null;
        foreach ($imageList as $img) {
            if (($img['size'] ?? '') === 'large') {
                $image = $img;
                break;
            }
        }
        $image = $image ?: end($imageList) ?: ['#text' => null];
        $this->imageUrl = $image['#text'];

        // Tags
        if (!empty($json['toptags']['tag'])) {
            foreach ($json['toptags']['tag'] as $tag) {
                $this->tags[] = $tag['name'] ?? '';
            }
        }

        // Tracks
        if (!empty($json['tracks']['track'])) {
            foreach ($json['tracks']['track'] as $track) {
                $this->tracks[] = [
                    'name' => $track['name'] ?? '',
                    'duration' => (int) ($track['duration'] ?? 0),
                    'url' => $track['url'] ?? '',
                    'rank' => isset($track['@attr']['rank']) ? (int) $track['@attr']['rank'] : null,
                    'artist' => $track['artist']['name'] ?? $this->artist,
                ];
            }
        }
    }
}

File: Mobile/media_tracker_mobile/lib/models/lastfm/lastfm_album.dart
class LastFmAlbum {
  final String name;
  final String artist;
  final String? imageUrl;
  final int? playCount;

  LastFmAlbum({
    required this.name,
    required this.artist,
    this.imageUrl,
    this.playCount,
  });

  factory LastFmAlbum.fromJson(Map<String, dynamic> json) {
    final imageList = json['image'] as List<dynamic>? ?? [];
    final image = imageList.lastWhere(
      (img) => img['#text'] != null && (img['#text'] as String).isNotEmpty,
      orElse: () => {'#text': null},
    );

    return LastFmAlbum(
      name: json['name'] ?? '',
      artist: json['artist']?['name'] ?? '',
      imageUrl: image['#text'],
      playCount: int.tryParse('${json['playcount'] ?? ''}'),
    );
  }
}
